### PR TITLE
Filesource compression support

### LINKED ISSUE #5245: Support compressed file sources
**Is your feature request related to a problem? Please describe.**
In order to develop ingestion pipelines for AWS infrastructure component logs (ALB, cloudfront...), I am testing my pipelines with files coming from this infra, which are compressed gzip files.

However, data-prepper does not support compression in the file source plugin, adding an extra step of decompressing the files manually.

In general this is a feature that would be welcomed as compressed log files are a common occurrence.

**Describe the solution you'd like**
the `file` source should have a `compression` field, set to `none` by default. When the plugin is started, the input stream that reads the file comes from the `DecompressionEngine` that corresponds to the `compression` field value.

**Describe alternatives you've considered (Optional)**
N/A

**Additional context**

### PR DESCRIPTION
**Pull request by joelmarty**
### Description
Add support for compressed files in file source.
 
### Issues Resolved
Fixes #5245
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
- [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/common/src/test/java/org/opensearch/dataprepper/plugins/source/file/FileSourceTests.java [hunk 1]
```diff
@@ -290,21 +321,18 @@ void setUp() {
 
         @Test
         void start_will_parse_codec_with_correct_inputStream() throws IOException {
-            createObjectUnderTest().start(buffer);
+            final FileInputStream decompressedStream = new FileInputStream(TEST_FILE_PATH_PLAIN);
```
[the remark this answers is not in the document]
Comment (dlvenable) [reply data-prepper-plugins/common/src/test/java/org/opensearch/dataprepper/plugins/source/file/FileSourceTests.java:324]: Rather than changing the whole test, could you just mock the `DecompressionEngine` to return exactly the bytes given?

Perhaps an alternative is to keep both tests.

**Review by dlvenable — CHANGES_REQUESTED** [review]
Thank you @joelmarty for this contribution! I have one comment on the tests, but the rest looks good.

File: NOTICE [hunk 2]
```diff
@@ -10,3 +10,7 @@ Foundation (http://www.apache.org/).
 
 This product includes software developed by
 Joda.org (http://www.joda.org/).
+
+This product includes software developed by
```
Comment (dlvenable) [inline NOTICE:14]: Is this a requirement from your company? We don't require contributors to add to this.

Comment (dlvenable) [reply NOTICE:14]: Please see if your company requires this. If not, it would be best to remove this change.

In the meantime, I'm also checking with the OpenSearch project to determine if this is an acceptable change or not.

File: data-prepper-plugins/common/src/test/java/org/opensearch/dataprepper/plugins/source/file/FileSourceTests.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (joelmarty) [reply data-prepper-plugins/common/src/test/java/org/opensearch/dataprepper/plugins/source/file/FileSourceTests.java:324]: I cannot keep the test as it is: the test relies on the previous implementation's behavior: it tries to read from the input stream created in `FileSource.CodecFileStrategy#start` but because this stream is created in a try-with-resource (see [FileSource.java L181](https://github.com/opensearch-project/data-prepper/pull/5255/files/4684124bbb14fd6d01e4933bebc53f58d8d58d67#diff-79b9db4732930a1271f9653a6dbf5141c204951d567927290333645b8212b926R181)), it is closed when the test tries to read it again on [FileSourceTest.java L303](https://github.com/opensearch-project/data-prepper/pull/5255/files/4684124bbb14fd6d01e4933bebc53f58d8d58d67#diff-614c993d9d093322214d71b14e196d14069d6a778280c1c08f93145f4a1a23c4L303).
In other words, the test used to work because the input stream *was never properly closed*. I could reintroduce what I believe is a bug to keep the test as-is, but IMHO it would make more sense to validate the test is equivalent to the previous implementation.

**Review by JonahCalvo — APPROVED** [review]

File: NOTICE (hunk 2 above)
Comment (joelmarty) [reply NOTICE:14]: Hi @dlvenable , any news on this ? The PR has been marked as accepted by @JonahCalvo but since there are failing build checks, I cannot merge it myself.

**Comment by joelmarty** [comment]
@JonahCalvo @dlvenable I resolved the merge conflict and reverted the changes in the NOTICE file.

**Comment by dlvenable** [comment]
@joelmarty , Sorry for the delay on this. I think I missed your most recent changes. We have merged this since you removed the NOTICE updates. Thank you.
